### PR TITLE
Add php 8 support

### LINKED ISSUE #5646: PHP 8.0 Roadmap
# Problem description

Wanted to have a look what would be needed to support PHP 8. At current state I must sadly tell it will not be possible without big effort to support sulu 1.6 and php 8.0 the first packages I checked with conflichts is the [DoctrineCacheBundle](https://github.com/doctrine/DoctrineCacheBundle) and [Puls00FFmpegBundle](https://github.com/pulse00/ffmpeg-bundle) both are abandoned and so will not merge any pull requests.

So I would think when not specially requested by a sponsor or partner we will focus PHP 8 support for the 2.x (current 2.1) Version.

# Proposed solution

## Conflicting packages

From the `composer why php` list:

 - [x] doctrine/orm (coming soon: https://twitter.com/beberlei/status/1329338553407578112)
 - [x] handcraftedinthealps/rest-routing-bundle https://github.com/handcraftedinthealps/RestRoutingBundle/pull/8
 - [x] handcraftedinthealps/zendsearch https://github.com/handcraftedinthealps/ZendSearch/pull/3
 - [x] friendsofsymfony/http-cache
 - [x] friendsofsymfony/http-cache-bundle
 - [x] friendsofsymfony/jsrouting-bundle
 - [x] symfony-cmf/slugifier-api https://github.com/symfony-cmf/slugifier-api/pull/6
 - [x] symfony-cmf/routing-bundle (only github actions: https://github.com/symfony-cmf/routing-bundle/pull/459)
    - [x] symfony/testing https://github.com/symfony-cmf/Testing/pull/208
 - [x] doctrine/phpcr-bundle https://github.com/doctrine/DoctrinePHPCRBundle/pull/355
 - [x] jackalope/jackalope-doctrine-dbal (only github actions https://github.com/jackalope/jackalope-doctrine-dbal/pull/378)
 - [x] jackalope/jackrabbit (only github actions https://github.com/jackalope/jackalope-jackrabbit/pull/165)
 - [x] phpcr/phpcr-migrations (https://github.com/phpcr/phpcr-migrations/pull/12)
   - [x] jackalope/jackalope-fs https://github.com/jackalope/jackalope-fs/pull/28
 - [x] league/flysystem-azure-blob-storage (doesnt conflict but need to be checked if works as has >=5.6 in it) https://github.com/sulu/sulu/pull/5751
 - [x] toflar/psr6-symfony-http-cache-store (needs update to 3.0 https://github.com/sulu/sulu/pull/5750) 
 - [x] phpcr/phpcr-shell https://github.com/phpcr/phpcr-shell/pull/209
 - [x] php-ffmpeg/php-ffmpeg (needs update to https://github.com/sulu/sulu/pull/5755)
 - [x] oro/doctrine-extensions (doesn't conflict but need to be checked if it works as  has >= 5.4 and no release since 2019) #5752
 - [x] matamo/device-detector #5643
      - [-] mustangostang/spyc (will be updated with #5643)
 - [x] layershifter/tld-database (will be removed in #5754)
 - [x] layershifter/tld-extract (will be removed in #5754)
 - [x] layershifter/tld-support (doesn't conflict but need to be checked as has >= 5.5 in it)  (will be removed in #5754)
 - [x] jms/serializer-bundle
 - [x] jms/serializer
 - [x] jms/metadata
 - [x] goodby/csv (need to be checked as >= 5.3) (hope it works because no release and no accepted PR since 2017)
 - [x] gedmo/doctrine-extensions (needs to be checked as >= 5.3) #5753
    - [-] behat/transliterator 
 - [x] JMS Serializer https://github.com/schmittjoh/JMSSerializerBundle/pull/830
 - [x] doctrine/* packages Upgrade https://github.com/sulu/sulu/pull/5774
 - [x] ~~php-http/guzzle6-adapter https://github.com/php-http/guzzle6-adapter/pull/86~~ Requirement removed in: https://github.com/sulu/sulu/pull/5811
 - [x] stof/doctrine-extensions-bundle replaced with antishov/doctrine-extensions-bundle https://github.com/sulu/sulu/pull/5814
 - [x] rokka-io/imagine-vips https://github.com/rokka-io/imagine-vips/pull/17 (requires release)
 - [x] Prepare PHP 8 compatibility https://github.com/sulu/sulu/pull/5813
 - [x] Add PHP 8 compatibility https://github.com/sulu/sulu/pull/5757
 
 The following should work but are not tested yet with PHP 8:
 
 - [ ] imagine/imagine (need to be checked as >= 5.3) https://github.com/avalanche123/Imagine/pull/758
 - [ ] superbalist/flysystem-google-storage (doesnt conflict but need to be checked if works as has >=5.5 and no release since 2019) https://github.com/Superbalist/flysystem-google-cloud-storage/pull/119
     - [-] google/cloud-core (same)
 - [ ] league/flysystem-aws-s3-v3 (need to be checked) https://github.com/thephpleague/flysystem-aws-s3-v3/pull/238
      - [-] aws/aws-sdk-php

 ## Check compatibility of the bundles:
 
  - [x] https://github.com/sulu/web-twig/pull/32
  - [x] https://github.com/sulu/SuluAutomationBundle/pull/48
  - [x] https://github.com/sulu/SuluFormBundle/pull/280
  - [x] https://github.com/sulu/SuluCommunityBundle/pull/141
  - [x] https://github.com/sulu/SuluRedirectBundle/pull/62
  - [x] https://github.com/sulu/SuluHeadlessBundle/pull/72
  - [x] https://github.com/sulu/SuluContentBundle/pull/187
  - [x] https://github.com/sulu/SuluArticleBundle/pull/546
      - [x] handcraftedinthealps/elasticsearch-dsl
      - [x] handcraftedinthealps/elasticsearch-bundle

## Known Issues

The following issues are not blocking but need to be keep in mind before upgrading a project to PHP 8:

 - [ ] https://github.com/doctrine/migrations/issues/1104 (Problems with transactions) (Workaround: using isTransactional false)
     - [ ] https://github.com/doctrine/dbal/pull/4481
 - [ ] https://github.com/Imagick/imagick/issues/358 (No PHP 8 release yet) (Workaround: Compiling Imagick yourself or use GD or Vips adapter)
 - [ ] https://github.com/deployphp/deployer/commit/042d1006906002ef580e5453894c4fd47180701e (No stable PHP 8 release yet)

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A
| --- | ---
| Bug fix? | no 
| New feature? |  yes
| BC breaks? | no 
| Deprecations? | no 
| Fixed tickets | fixes #5646 
| Related issues/PRs | # 
| License | MIT
| Documentation PR | sulu/sulu-docs# 

#### What's in this PR?

Add php 8 support

#### Why?

Support newest php version so our community can use it

#### Requires

 - [x]  phpcr/phpcr-migrations https://github.com/phpcr/phpcr-migrations/pull/12
  - [x] Upgrade Doctrine Packages https://github.com/sulu/sulu/pull/5774

### CODE REVIEW

File: .github/workflows/test-application.yaml [hunk 1]
```diff
@@ -104,6 +104,20 @@ jobs:
...
+                          SYMFONY_DEPRECATIONS_HELPER: weak
+                          PHPCR_TRANSPORT: doctrinedbal
+                          DATABASE_URL: mysql://root:root@127.0.0.1:3306/sulu_test?serverVersion=5.7
+                          DATABASE_CHARSET: utf8mb4
+                          DATABASE_COLLATE: utf8mb4_unicode_ci
+
+                    - php-version: '8.0'
+                      database: mysql
+                      phpcr-transport: doctrinedbal
+                      composer-flags: '--prefer-dist --no-interaction --ignore-platform-req="php"' # --ignore-platform-req should be removed
```
Comment (alexander-schranz) [inline .github/workflows/test-application.yaml:118]:
```
--ignore-platform-req="php"
```

should be removed when the linked dependencies of #5646 are released

File: src/Sulu/Bundle/RouteBundle/Generator/SymfonyExpressionTokenProvider.php [hunk 2]
```diff
@@ -54,7 +54,7 @@ public function provide($entity, $name)
         $locale = $this->translator->getLocale();
 
         try {
-            if (\method_exists($entity, 'getLocale')) {
+            if (\is_object($entity) && \method_exists($entity, 'getLocale')) {
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/RouteBundle/Generator/SymfonyExpressionTokenProvider.php]: See: https://3v4l.org/RZo8X

File: src/Sulu/Component/Rest/RestControllerTrait.php [hunk 3]
```diff
@@ -214,7 +214,7 @@ protected function processPut($entities, $requestEntities, $deleteCallback, $upd
         // FIXME: this is just a hack to avoid relations that start with index != 0
         // FIXME: otherwise deserialization process will parse relations as object instead of an array
         // reindex entities
-        if (\count($entities) > 0 && \method_exists($entities, 'getValues')) {
+        if (\count($entities) > 0 && \is_object($entities) && \method_exists($entities, 'getValues')) {
```
Comment (alexander-schranz) [inline src/Sulu/Component/Rest/RestControllerTrait.php]: See https://3v4l.org/RZo8X

File: src/Sulu/Component/Util/SuluNodeHelper.php [hunk 4]
```diff
@@ -250,7 +250,8 @@ public function extractSnippetTypeFromPath($path)
         $newPath = PathHelper::getParentPath($path);
         $newPath = \substr($newPath, \strlen($snippetsPath));
 
-        if (false === $newPath) {
+        // $newPath can be false or empty because of return difference of substr depending on php version (<= 7.4, 8.0)
+        if (!$newPath) {
```
Comment (alexander-schranz) [inline src/Sulu/Component/Util/SuluNodeHelper.php]: See https://3v4l.org/4uZIc

File: src/Sulu/Component/Util/Tests/Unit/SortUtilsTest.php [hunk 5]
```diff
@@ -105,16 +105,26 @@ public function provideSortObjects()
                 '[foo]', ['2', '1', '3'],
             ],
 
-            // multi dimensional array missing key
+            // multi dimensional array missing key asc
```
Comment (alexander-schranz) [inline src/Sulu/Component/Util/Tests/Unit/SortUtilsTest.php]: two null values are different, to test the order I created a asc and desc test based on object:

See also https://3v4l.org/LsUj7

File: src/Sulu/Component/Util/Tests/Unit/SuluNodeHelperTest.php [hunk 6]
```diff
@@ -179,6 +179,7 @@ public function provideExtractSnippetTypeFromPath()
         return [
             ['/cmf/snippets/foobar/snippet1', 'foobar'],
             ['/cmf/snippets/bar-foo/snippet2', 'bar-foo'],
+            ['/cmf', null, false],
```
Comment (alexander-schranz) [inline src/Sulu/Component/Util/Tests/Unit/SuluNodeHelperTest.php]: just additonal test case to be sure that 7.4 and 8.0 result in the same here

**Comment by alexander-schranz** [comment]
@sulu/core-team Which PHP Versions should we test. The current setup is:

 - Github: PHP 7.2 (postgres, jackrabbit) (prefer-lowest)
 - Github:  PHP 7.4 (mysql 5.7, doctrinedbal)
 - AppVeyor:  PHP 7.2 (mysql 8, doctrinedbal)
 
 I would maybe go here with:

 - Github: PHP 7.2 (postgres, jackrabbit)  (prefer-lowest)
 - Github:  PHP 8.0 (mysql 5.7, doctrinedbal)
 - AppVeyor:  PHP 7.4 (mysql 8, doctrinedbal)

**Comment by luca-rath** [comment]
I would go with:

 - Github: PHP 7.4 (postgres, jackrabbit)
 - Github:  PHP 8.0 (mysql 8, doctrinedbal)
 - AppVeyor:  PHP 7.2 (mysql 5.7, doctrinedbal)
 
to speed up github actions, so contributors have faster ci results

**Comment by alexander-schranz** [comment]
@luca-rath running prefer-lowest on windows is not a good option will take there a lot more time as it current does.

> to speed up github actions, so contributors have faster ci results

So for faster result PHP 7.2 and 8.0 should be keeped on github action, as this is the most needed requirement that a PR is being merged. Just more ask here if we should add additional 7.4 on Github Action or move it to the Appveyor build.

**Comment by luca-rath** [comment]
I guess this `PHP 8.0 (mysql 8, doctrinedbal)` should be the fastest job ... i think this job should run on github actions with linting enabled ... I don't care about the other jobs

**Comment by luca-rath** [comment]
I don't think we need a php7.4 on github actions then

**Comment by Edvinas9** [comment]
🎉 `phpcr/phpcr-migrations` already released version **1.3.0** with **PHP 8 support**: https://github.com/phpcr/phpcr-migrations/releases/tag/1.3.0

**Comment by alexander-schranz** [comment]
@Edvinas9 Thx we will merge as next step: https://github.com/sulu/sulu/pull/5774 and I will then test it against a Production Project to be sure that it works. Maybe need to upgrade to sulu packages also to support newer doctrine version.

**Review by niklasnatter — APPROVED** [review]

**Pull request merged**